infra/lib/api-stack.ts

```diff
@@ -496,6 +465,155 @@ export class APIStack extends Stack {
     garminResource.addMethod("ANY", new apig.LambdaIntegration(garminLambda));
   }
 
+  private setupTokenAuthLambda(dynamoDBTokenTable: dynamodb.Table): apig.RequestAuthorizer {
+    const tokenAuthLambda = new lambda_node.NodejsFunction(this, "APITokenAuthLambda", {
+      runtime: lambda.Runtime.NODEJS_16_X,
+      entry: "../api/lambdas/token-auth/index.js",
+      environment: {
+        TOKEN_TABLE_NAME: dynamoDBTokenTable.tableName,
+      },
+    });
+    addErrorAlarmToLambdaFunc(this, tokenAuthLambda, "TokenAuthFunctionAlarm");
+
+    const tokenAuth = new apig.RequestAuthorizer(this, "APITokenAuth", {
+      handler: tokenAuthLambda,
+      identitySources: ["method.request.querystring.state"],
+      // todo: instead of removing caching, investigate explicitly listing
+      //        the permitted methods in the lambda: "Resource: event.methodArn"
+      //
+      // see: https://forum.serverless.com/t/rest-api-with-custom-authorizer-how-are-you-dealing-with-authorization-and-policy-cache/3310
+      resultsCacheTtl: Duration.minutes(0),
+    });
+    tokenAuthLambda.role && dynamoDBTokenTable.grantReadData(tokenAuthLambda.role);
+
+    return tokenAuth;
+  }
+
+  private setupAPIGWApiTokenResource(
+    stackId: string,
+    api: apig.RestApi,
+    link: apig.VpcLink,
+    authorizer: apig.RequestAuthorizer,
+    serverAddress: string
+  ): apig.Resource {
+    const apiTokenResource = api.root.addResource("token");
+    const tokenProxy = new apig.ProxyResource(this, `${stackId}/token/Proxy`, {
+      parent: apiTokenResource,
+      anyMethod: false,
+    });
+    const integrationToken = new apig.Integration({
+      type: apig.IntegrationType.HTTP_PROXY,
+      options: {
+        connectionType: apig.ConnectionType.VPC_LINK,
+        vpcLink: link,
+        requestParameters: {
+          "integration.request.path.proxy": "method.request.path.proxy",
+          "integration.request.header.api-token": "context.authorizer.api-token",
+          "integration.request.header.cxId": "context.authorizer.cxId",
+          "integration.request.header.userId": "context.authorizer.userId",
+        },
+      },
+      integrationHttpMethod: "ANY",
+      uri: `http://${serverAddress}/{proxy}`,
+    });
+    tokenProxy.addMethod("ANY", integrationToken, {
+      requestParameters: {
+        "method.request.path.proxy": true,
+      },
+      authorizer,
+    });
+    return apiTokenResource;
+  }
+
+  private setupOAuthUserPool(): cognito.IUserPool {
+    const userPool = new cognito.UserPool(this, "oauth-client-secret-user-pool", {
+      accountRecovery: cognito.AccountRecovery.EMAIL_ONLY,
+      removalPolicy: RemovalPolicy.DESTROY,
+    });
+    // TODO make this a custom domain
+    userPool.addDomain("metriport-cognito-domain", {
+      cognitoDomain: {
+        domainPrefix: "metriport", // TODO make this dynamic/config
```

Comment:
This should be updated when we create a custom domain for cognito, as part of the ticket (there's a TODO for it): https://github.com/metriport/metriport-internal/issues/228